### PR TITLE
[AC97] Add Open Source AC'97 Driver

### PR DESCRIPTION
**Pull request by Obaid51**
This PR imports the AC'97 Driver that was originally modified by Michael Stamper. His PR can be found at https://github.com/reactos/reactos/pull/3141
I am not sure about zero3k which you are referring to, but I was hired by some guy on a freelancing platform for some coding. Also as this is working without installing the audio driver explicitly, I guess it's good for the pull request.
Proof of working can be found at the following video link
https://youtu.be/Pv3q_ZLAYt8

### CODE REVIEW

**Comment by gonzoMD** [comment]
I'll close this one for several reasons:

There were several attempts to add this in the past by a guy who is not very well accepted by the community.
I suspect that it's either you (hello zero3k), or that, at least, he pushed you to send this pull request (he did that in the past).
Also there is no description, no tests, no prove that it works.
If it has nothing to do with zero3k, feel free to get in touch with us, but at this state, we will not accept it.

**Comment by JoachimHenze** [comment]
When running that SuperDX ball from the attached video with that new driver, and then exiting SuperDX-ball, will that trigger a heap assertion @Obaid51, or "03k", or "Bryan Kirk"?

**Comment by gonzoMD** [comment]
OK. I'll reopen it for now. Some people gave careful positive feedback. Give it some time because this has to be tested very well because of known bugs, at least with the proprietary ac97 driver. If the improvement is better than the drawbacks, there is a chance.

**Comment by Obaid51** [comment]
Super-DX Ball does not cause a heap assertion for me after I have closed it with the driver included in this PR being installed. 
The patch at https://jira.reactos.org/secure/attachment/61263/NTOS%20MM%20-%20undo%20%09c7e090.diff should help make sure it doesn't happen for anyone that is experiencing it.

**Comment by ROCKNROLLKID** [comment]
I am not a developer, however before any work should be done on this, can you tell us who actually hired you. I don't think anyone from the actual team hired anyone, so this is likely someone else who did. The reason for our concerned is because the person (zero3K/Bryan) has done this before, and it ended up being a scam. See here for the other pr that he "hired" and ended up being a scam: https://github.com/reactos/reactos/pull/3653

**Comment by Obaid51** [comment]
@ROCKNROLLKID Yes, Bryan Kirk (Zero3K) hired me via Upwork.com. I was informed by other developers about his past experiences in the chat room. So I confronted him and then paid the money he agreed to pay in the job post.

**Comment by jimtabor** [comment]
I have tested and built the original source and it crashes.....

Fix the crash and we can talk.

**Comment by binarymaster** [comment]
@jimtabor from what I remember you got the crash in QEMU 6.x.x, can you please share your configuration and exact QEMU version?

That could help to reproduce the crash, so @Obaid51 could fix it.

**Comment by sskras** [comment]
@gonzoMD commented [3 days ago](#issuecomment-1004597470)

> Also there is no description, no tests, no prove that it works.

By the way, speaking of tests: is there any documentation about extending the existing ReactOS test infrastructure?

When to use the `apitests`, the `tests` or the `winetests` in case of regular APIs?
And this case: the `drivers`, the `kmtests` or the `win32`?

I found only [Writing tests](https://reactos.org/wiki/Writing_tests) page on wiki. Tried reading it, but it looks both too bloated and non-uniform (in some places too dense, in some too scarce) to me.

Eg. it still unclear in what dir in the tree should I put these `.xml` files.

**Comment by learn-more** [comment]
> @gonzoMD commented [3 days ago](#issuecomment-1004597470)
> 
> > Also there is no description, no tests, no prove that it works.
> 
> By the way, speaking of tests: is there any documentation about extending the existing ReactOS test infrastructure?
> 
> When to use the `apitests`, the `tests` or the `winetests` in case of regular APIs? And this case: the `drivers`, the `kmtests` or the `win32`?
> 
> I found only [Writing tests](https://reactos.org/wiki/Writing_tests) page on wiki. Tried reading it, but it looks both too bloated and non-uniform (in some places too dense, in some too scarce) to me.
> 
> Eg. it still unclear in what dir in the tree should I put these `.xml` files.

This page is completely outdated.
Apitests are our own tests, winetests are apitests from wine.
kmtests are driver tests

**Comment by sskras** [comment]
@learn-more commented [16 minutes ago](#issuecomment-1007300514):

> kmtests are driver tests

Thanks.

So this PR will need test cases, which should be added into `kmtests\ac97`, right?

Or maybe some generic `kmtests\audio` dir?
(as there is common part of audio subsystem that's involved into running ac97, no?)

Maybe `kmtests\wdm\audio`? (as this is WDM driver)

Maybe even `kmtests\wdm\audio\legacy\wdmaud` which would reflect [drivers\wdm\audio\legacy\wdmaud](https://github.com/reactos/reactos/tree/master/drivers/wdm/audio/legacy/wdmaud) path I just found in the repo?

Or did I get the comment from @gonzoMD wrong?

File: drivers/wdm/audio/drivers/ac97/ac97.inf [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+ICHPin.Front            ="Front Speaker"
+
+[Strings.0407]
+ProviderName="FooProviderName"
+MfgName="Intel"
+DiskDescription="Diskette f�r AC'97 WDM Treiberbeispiel"
+ac97_AA.DeviceDesc="AC'97 Treiberbeispiel f�r Intel(r) 82801AA Kontroller"
+ac97_AB.DeviceDesc="AC'97 Treiberbeispiel f�r Intel(r) 82801AB Kontroller"
+ac97_BA.DeviceDesc="AC'97 Treiberbeispiel f�r Intel(r) 82801BA/BAM Controller"
+ac97.DeviceDesc="AC'97 Treiberbeispiel f�r Intel(r) 82801 Kontroller"
```
Comment (binarymaster) [inline drivers/wdm/audio/drivers/ac97/ac97.inf:400]:
```suggestion
ac97_AA.DeviceDesc="Intel 82801AA AC'97 Audiocontroller"
ac97_AB.DeviceDesc="Intel 82801AB AC'97 Audiocontroller"
ac97_BA.DeviceDesc="Intel 82801BA/BAM AC'97 Audiocontroller"
ac97.DeviceDesc="Intel 82801 AC'97 Audiocontroller"
```

File: sdk/include/ddk/portcls.h [hunk 2]
```diff
@@ -704,6 +704,16 @@ typedef IUnregisterPhysicalConnection *PUNREGISTERPHYSICALCONNECTION;
...
+/* C++ ABI HACK: IDmaChannel::PhysicalAddress */
+#if defined(__cplusplus) && !defined(_MSC_VER)
+#define DEFINE_ABSTRACT_DMACHANNEL_PhysicalAddress \
+    STDMETHOD_(PHYSICAL_ADDRESS*, PhysicalAddress)( THIS_ PHYSICAL_ADDRESS* pRet ) PURE; \
+    PHYSICAL_ADDRESS PhysicalAddress() { PHYSICAL_ADDRESS tmp; PhysicalAddress(&tmp); return tmp; }
+#else
+#define DEFINE_ABSTRACT_DMACHANNEL_PhysicalAddress \
+    STDMETHOD_(PHYSICAL_ADDRESS, PhysicalAddress)( THIS ) PURE
+#endif
+
```
Comment (binarymaster) [inline sdk/include/ddk/portcls.h:716]: All these changes outside the driver also must go by a separate commit(s).

**Review by binarymaster — CHANGES_REQUESTED** [review]

File: drivers/wdm/audio/drivers/ac97/license.txt [hunk 3]
```diff
@@ -0,0 +1,21 @@
+    MIT License
```
Comment (Extravert-ir) [inline drivers/wdm/audio/drivers/ac97/license.txt:1]: Please, remove extra spaces in the beginning (all lines)

File: sdk/include/psdk/ksmedia.h [hunk 4]
```diff
@@ -508,6 +508,106 @@ DEFINE_GUIDSTRUCT("DFF21CE6-F70F-11D0-B917-00A0C9223196", KSNODETYPE_COMMUNICATI
 DEFINE_GUIDSTRUCT("DFF21CE7-F70F-11D0-B917-00A0C9223196", KSNODETYPE_LOW_FREQUENCY_EFFECTS_SPEAKER);
 #define KSNODETYPE_LOW_FREQUENCY_EFFECTS_SPEAKER DEFINE_GUIDNAMED(KSNODETYPE_LOW_FREQUENCY_EFFECTS_SPEAKER)
 
+#define STATIC_KSAUDFNAME_PC_SPEAKER_VOLUME\
```
Comment (Extravert-ir) [inline sdk/include/psdk/ksmedia.h:511]: Either move the change to a separate commit, or move it to a separate PR

File: drivers/wdm/audio/drivers/ac97/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    wavecyclicminiport.cpp
+    wavecyclicstream.cpp
+)
+
+add_library(ac97 MODULE
+    ${SOURCE}
+    ac97.rc)
+
+target_link_libraries(ac97 stdunk libcntpr uuid)
+set_module_type(ac97 wdmdriver UNICODE)
```
Comment (Extravert-ir) [inline drivers/wdm/audio/drivers/ac97/CMakeLists.txt:23]:
```suggestion
set_module_type(ac97 wdmdriver)
```

File: drivers/wdm/audio/drivers/ac97/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+add_importlibs(ac97 portcls hal ntoskrnl)
+
+if(USE_CLANG_CL OR (NOT MSVC))
```
Comment (Extravert-ir) [inline drivers/wdm/audio/drivers/ac97/CMakeLists.txt:26]:
```suggestion
if(CMAKE_C_COMPILER_ID STREQUAL "GNU" OR CMAKE_C_COMPILER_ID STREQUAL "Clang")
```

File: drivers/wdm/audio/drivers/ac97/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    target_compile_options(ac97 PRIVATE -Wno-write-strings -Wno-switch)
+endif()
+
+add_cd_file(TARGET ac97 DESTINATION reactos/system32/drivers FOR all)
+add_cd_file(FILE "${CMAKE_CURRENT_SOURCE_DIR}/ac97.inf" DESTINATION reactos/inf FOR all)
```
Comment (Extravert-ir) [inline drivers/wdm/audio/drivers/ac97/CMakeLists.txt:31]:
```suggestion
add_driver_inf(ac97 ac97.inf)
```

File: drivers/wdm/audio/drivers/ac97/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    miniport.cpp
+    stream.cpp
+    stream2.cpp
+    rtminiport.cpp
+    rtstream.cpp
+    wavepciminiport.cpp
+    wavepcistream.cpp
+    wavecyclicminiport.cpp
+    wavecyclicstream.cpp
+)
```
Comment (Extravert-ir) [inline drivers/wdm/audio/drivers/ac97/CMakeLists.txt:16]:
```suggestion
    wavecyclicstream.cpp)
```

**Review by Extravert-ir — CHANGES_REQUESTED** [review]
Information about the driver needs to be added to https://github.com/reactos/reactos/blob/master/media/doc/3rd%20Party%20Files.txt file.
Now please address remaining remarks and we're good to go.

**Comment by Vayne2500** [comment]
How do I fix the duplicate commits?

**Comment by Vayne2500** [comment]
I have done all of the requested changes under my Zero3K account.

**Pull request closed without merging**